Better Explain State Across All Develop (#1251)
near/docs: docs/2.develop/contracts/anatomy.md

---
id: anatomy

> title: The Contract Class

Reviewer: nice name change

<CodeTabs>
  <Language value="🌐 JavaScript" language="ts">
    <Github fname="contract.ts"
      url="https://github.com/near-examples/donation-js/blob/master/contract/src/contract.ts"
      start="5" end="9" />
  </Language>
  <Language value="🦀 Rust" language="rust">
    <Github fname="lib.rs"
      url="https://github.com/near-examples/donation-rust/blob/main/contract/src/lib.rs"
      start="7" end="12" />
  </Language>
</CodeTabs>

Under the hood, the `NEAR Bindgen` decorator/macro traverses the class, generating the necessary code to:

1. Transform the code into a valid NEAR contract.
2. Expose public methods, so they can be called externally.
3. Serialize objects for internal storage and communication with external actors.

<hr class="subsection" />

### The State
Each account has its own state (storage), which **only they can modify** but [anyone can see](../../4.tools/cli.md#near-view-state-near-view-state).

The state is defined and modified through the **main class' attributes**.

Contracts [**pay for their storage**](./storage.md#storage-cost) by locking part of their balance. Currently it costs **~1 Ⓝ** to store **100kb**

> :::info Key-Value Storage
> The contract actually uses a key-value storage to persist values. This however is abstracted from you
> by the SDK through [serialization](./serialization.md).
> :::

Reviewer 2: ```suggestion
:::info Key-Value Storage

The contract actually uses a `key-value` storage to persist values. This however is abstracted from you
by the SDK through [serialization](./serialization.md).

:::
```

:::tip Prefer SDK Collections
When defining attributes, **always prefer [SDK collections](./storage.md)** over native ones, since they are optimized for [serialization](./serialization.md).
:::

---

## Initializing the State

> There are two ways to initialize the accounts's state, and they an co-exist:

Reviewer 2: ```suggestion
There are two ways to initialize the account's state, and they can co-exist:
```

1. An **initialization method** that receives the attributes needed for the state

> 2. A **default state**, which will be used until `init` in invoked, or a method writes into the state

Reviewer 2: ```suggestion
2. A **default state**, which will be used until `init` is invoked, or a method writes into the state
```

<hr class="subsection" />

### Initialization Method
To define an initialization method simply decorate it with the [initialization macro](#decorators--macros).

> The method will now be able to define the initial values of the state, raising an error if invoked

Reviewer 2: ```suggestion
The method will now be able to define the initial state's values, raising an error if invoked
```

          url="https://github.com/near-examples/donation-js/blob/master/contract/src/contract.ts"
          start="10" end="13" />

:::info
To make the initialization mandatory use `@NearBindgen({requireInit: true})`
:::

:::caution

> In Javascript you **must always** define a [default state](#default-state)

Reviewer 2: ```suggestion
In JavaScript you **must always** define a [default state](#default-state)
```

          url="https://github.com/near-examples/donation-js/blob/master/contract/src/contract.ts"
          start="6" end="8" />

  🌐 In JavaScript, the default state is defined by the initialization parameters in the class definition.

:::caution

> In Javascript you **must always** give a value to **all parameters** of the class.

Reviewer: It would be nice for this to explain shortly why this is necessary, and link to storage/state section.
```suggestion
In Javascript you **must always** give a value to **all parameters** of the class. This ensures that values, when deserialized are attributed to their intended types ([read more on state and deserialization](./serialization.md)).
```

be called for free by everyone, **without needing** a NEAR account.
- **Call Methods**: They can mutate the state and perform [actions](./actions.md) such
as calling other contracts.

:::caution
By default `view` methods have `200TGas` to execute, to increase this you can simple invoke them
as `call` methods
:::

<hr class="subsection" />

### Private Methods
Sometimes you will want some methods to remain public, but only be callable by the contract's
account. Such is the case for example of [cross-contract callbacks](./crosscontract.md#callback-method).

For this, you can use the `private` macro/decorator.

<Tabs className="language-tabs" groupId="code-tabs">
  <TabItem value="🌐 JavaScript">

  ```ts
  @call({privateFunction: true})
  callback( ... ){
    // this method can only be called by the contract's account
  }
  ```

  </TabItem>
  <TabItem value="🦀 Rust">

  ```rust
  #[private]
  pub fn callback(&mut self, ... ){
    // this method can only be called by the contract's account
  }
  ```

  </TabItem>
</Tabs>

<hr class="subsection" />

### Payable Methods
By default **all methods panic** if a user **attaches money** while calling them. To enable a

> method to receive money use the payable decorator.
> <Tabs className="language-tabs" groupId="code-tabs">

Reviewer 2: ```suggestion
method to receive money use the payable decorator.

<Tabs className="language-tabs" groupId="code-tabs">
```
